pkg/virt-handler/heartbeat/multipath-socket.go

```diff
@@ -0,0 +1,77 @@
+package heartbeat
+
+import (
+	"errors"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/storage/reservation"
+)
+
+const (
+	runDir              = "/proc/1/root/run"
+	socket              = "multipathd.socket"
+	multipathSocketPath = runDir + "/" + socket
+)
+
+type MonitorMultipathSocket struct {
+	w *fsnotify.Watcher
+}
+
+func NewMonitorMultipathSocket() *MonitorMultipathSocket {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed creating watcher for multipath socket")
+	}
+	w.Add(runDir)
+	return &MonitorMultipathSocket{w: w}
+}
+
+func (m *MonitorMultipathSocket) Run() error {
+	dir := reservation.GetPrHelperHostSocketDir()
+	path := filepath.Join(dir, socket)
+
+	if _, err := os.Stat(dir); errors.Is(err, os.ErrNotExist) {
+		if err := os.Mkdir(dir, 0700); err != nil {
+			log.Log.Reason(err).Errorf("failed creating persistent reservation directory %s", dir)
+		}
+	}
+	if _, err := os.Stat(multipathSocketPath); !errors.Is(err, os.ErrNotExist) {
+		// Remove the hardlink in case it already exists and recreates it
+		os.Remove(path)
+		if out, err := exec.Command("ln", multipathSocketPath, path).CombinedOutput(); err != nil {
+			log.Log.Reason(err).Errorf("creating link %s for multipath socket: %s", path, string(out))
+		}
+	}
+	for {
+		select {
+		case event := <-m.w.Events:
+			// Filter the event by path
+			if event.Name != multipathSocketPath {
+				continue
+			}
+			if event.Op&fsnotify.Remove == fsnotify.Remove {
+				log.Log.Infof("Removed link %s", path)
+				os.Remove(path)
+				continue
+			}
+			if _, err := os.Stat(path); errors.Is(err, os.ErrNotExist) {
+				if out, err := exec.Command("ln", multipathSocketPath, path).CombinedOutput(); err != nil {
```

Comment:
I wonder, in which cases this can fail. maybe we should do more than just log if this fails multiple times?

Reply:
I think it can fail if we have a permission issues. In all cases, I don't think we can do much. I wanted also to keep this self-contained so I would prefer to handle the error in this context. If you have any suggestions of possible issue, we can try to resolve them here

Reply:
maybe an event if it fails on the same error several times? This error may not be visible enough for the admin. It will of course pollute the logs :)